rfcs/0000-poa-testnet/0000-poa-testnet.md

```diff
@@ -0,0 +1,183 @@
+* The network cannot be halted or concerned by minority malicious validators, malicious validators should be eventually evicted by majority honest validators.
+* The other part of the network should be the same as the mainnet(unless block header may need some extract context to do verification).
+
+## POA protocol
+
+We define the following variables:
+
+* `VALIDATOR_COUNT` - Number of current validators, this value changed due to new validators join or old validator leaves.
+* `ATTEST_INTERVAL` - A validator cannot attest two blocks within `ATTEST_INTERVAL` number. For example, a validator who attest block (6) must wait for at least `ATTEST_INTERVAL` blocks to do next attest: block (6 + `ATTEST_INTERVAL` + 1). Notice when the `VALIDATOR_COUNT` <= `ATTEST_INTERVAL`, the POA testnet will stuck forever due to no validators can attest a new block.
+* `BLOCK_INTERVAL` - the interval of blocks, set to 8 seconds.
+* `VOTE_LIMIT` - The least votes to make a new validator join or to evict an old validator, should be at least `VALIDATOR_COUNT / 2 + 1`.
+
+### attest a new block
+
+Instead of POW mining, a list of validators plays the role of producing blocks in POA testnet. Validators work in a round-robin style; in each block height, there exists a corresponded validator that attests to the new block by sign the block with its private key. For convenient, we call the validator who attest a block an **attester** to distinguish it from other validators.
+
+The community chooses the initial validator list of the POA network by some off-chain governance mechanism, which does not cover in this proposal. At this moment, we can ignore the possible updating of the validator list; our protocol is simple enough works for both dynamic and fixed validators.
+
+The protocol uses `n % VALIDATOR_COUNT` as the index to choose the attester from validators; each validator checks the index and decides whether to attest a new block itself or wait for another attester to produce a new block.
+
+However, an in-turned attester may fail to produce a block due to network error or other reasons; in this case, other validators must produce the new block to make the POA testnet continue.
+
+Validators can use a simple strategy:
+
+1. For block height `n` a validator checks `n % VALIDAROR_COUNT`.
+2. If `INDEX == n % VALIDATOR_COUNT`, which means the validator is in it's turn to attests block `n`. The validator should wait for `BLOCK_INTERVAL` seconds then attests the block `n` with difficulty set to `2`.
+3. If `INDEX != n % VALIDATOR_COUNT`, which means the validator is not in it's turn to attests block `n`. The validator should wait for `BLOCK_INTERVAL + rand(VALIDATOR_COUNT) * 0.5` seconds to wait for another attester to produce a new block. If there are no new block produced during the time, the validator should attest a new block with difficulty set to `1`.
```

Comment:
Could the result of `BLOCK_INTERVAL + rand(VALIDATOR_COUNT) * 0.5` be `BLOCK_INTERVAL`? Or very close to `BLOCK_INTERVAL`? If so a validator and attester may generate a new block at nearly the same time, and the validator's block arrives other nodes earlier than the attester's. This could cause a lot of 1-block fork switch on testnet.

I suggest adding a buffer to validator's wait period, e.g. `BLOCK_INTERVAL * 2 + rand(VALIDATOR_COUNT) * 0.5`.

Reply:
We could add another variable `BLOCK_TIMEOUT`, set a value that slightly greater than `BLOCK_INTERVAL` like `10s`.

Every validator who not in its turn waits for `BLOCK_TIMEOUT + rand(VALIDATOR_COUNT) * 0.5` seconds to produce a block, this reduces the possibilities that two validators attest in the same time.

In fact, even there are two blocks produced at the same time it's won't hurt; the block that attested by an in-turn validator always has higher `total difficulty`; even if two blocks have same `total difficulty`, the next in turned validator can decide the main chain, the chain selection is just like the POW.